fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java

```diff
@@ -0,0 +1,141 @@
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class ResourceQueue {
+    private static final AtomicLong QUEUE_ID = new AtomicLong(0);
+
+    private final long id;
+    private final String name;
+    private final ResourceQueueConfig config;
+    private final MatchingPolicy policy;
+    private final AsyncJobManager jobMgr;
+
+    // jobId -> job
+    private final Map<Long, AsyncJob> allJobs;
+    private final BlockingQueue<AsyncJob> pendingQueue;
+    private final Map<Long, AsyncJob> runningJobs;
+    private final Map<Long, AsyncJob> finishedJobs;
+    private final Semaphore runningSemaphore;
+
+
+    public ResourceQueue(String name, AsyncJobManager jobMgr, ResourceQueueConfig config, MatchingPolicy policy) {
+        this.id = QUEUE_ID.getAndIncrement();
+        this.name = name;
+        this.jobMgr = jobMgr;
+        this.config = config;
+        this.policy = policy;
+
+        allJobs = new HashMap<>();
+        pendingQueue = new ArrayBlockingQueue<>(config.max_queue_size());
+        runningJobs = new HashMap<>();
+        finishedJobs = new HashMap<>();
+        runningSemaphore = new Semaphore(config.max_concurrency());
+    }
+
+    public boolean match(AsyncJob job) {
+        return policy.match(job.getUser());
+    }
+
+    public synchronized boolean addJob(AsyncJob job) throws UserException {
+        if (allJobs.containsKey(job.jobId())) {
+            job.setJobStatus(JobStatus.CANCELED);
+            throw new UserException(String.format("Submit the same job(jobId=%d)", job.jobId()));
+        }
+        job.setJobStatus(JobStatus.PENDING);
+        if (pendingQueue.size() < config.max_queue_size() && pendingQueue.offer(job)) {
+            allJobs.put(job.jobId(), job);
+        }
+        job.setJobStatus(JobStatus.CANCELED);
+        return false;
+    }
+
+    public AsyncJob cancelJob(long jobId) {
+        AsyncJob job = allJobs.get(jobId);
+        if (job != null) {
+            synchronized (this) {
+                job = allJobs.get(jobId);
+            }
+        }
+        if (job != null) {
+            // cancel should be Idempotent
+            job.cancel();
+            synchronized (this) {
+                if (finishedJobs.putIfAbsent(jobId, job) == null) {
+                    if (runningJobs.remove(jobId) == null) {
+                        pendingQueue.remove(job);
+                    }
+                }
+            }
+        }
+        return job;
+    }
+
+    public ResourceQueueConfig getConfig() {
+        return config;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public boolean canAddJob() {
+        return pendingQueue.size() < config.max_queue_size();
+    }
+
+    public int numPendingJobs() {
+        return pendingQueue.size();
+    }
+
+    public int numRunningJobs() {
+        return runningJobs.size();
+    }
+
+    public long queueId() {
+        return id;
+    }
+
+    public void run() {
```

Comment:
I think we should use one thread to schedule jobs in all queue.

Reply:
done